### PR TITLE
Pylint

### PR DESCRIPTION
**Pull request by ryansanford**
Enable pylint checks via travis build. 

Do not merge until Travis build is successful. Establishing which messages to address vs. enable/disable will be discussed within this PR.

- Pylint is run after successful unit/integration tests
- Pylint configuration is contained in .pylintrc
- Configuration reference = https://docs.pylint.org/features.html
- Example of overriding check in source = https://docs.pylint.org/message-control.html

### CODE REVIEW

**Comment by ryansanford** [comment]
@gsfr Here's the breakout of failed checks. Looking for input on which checks can be disabled vs. resolved.

### Message errors by type
Check|Hits
------|--------
C0103:invalid-name|190
C0111:missing-docstring|185
C0113:unneeded-not|1
C0121:singleton-comparison|2
C0123:unidiomatic-typecheck|1
C0301:line-too-long|55
C0321:multiple-statements|2
C0325:superfluous-parens|1
C0326:bad-whitespace|184
C0330:bad-continuation|37
C0410:multiple-imports|1
C0411:wrong-import-order|52
E0203:access-member-before-definition|2
E0401:import-error|7
E0602:undefined-variable|21
E1101:no-member|23
E1120:no-value-for-parameter|4
E1121:too-many-function-args|1
R0201:no-self-use|10
R0204:redefined-variable-type|1
R0801:duplicate-code|4
R0902:too-many-instance-attributes|5
R0903:too-few-public-methods|4
R0911:too-many-return-statements|2
R0912:too-many-branches|5
R0913:too-many-arguments|7
R0914:too-many-locals|6
R0915:too-many-statements|2
W0102:dangerous-default-value|1
W0105:pointless-string-statement|2
W0106:expression-not-assigned|1
W0110:deprecated-lambda|2
W0120:useless-else-on-loop|4
W0141:bad-builtin|3
W0201:attribute-defined-outside-init|50
W0212:protected-access|10
W0221:arguments-differ|1
W0231:super-init-not-called|1
W0311:bad-indentation|2
W0403:relative-import|11
W0511:fixme|23
W0603:global-statement|1
W0611:unused-import|22
W0612:unused-variable|32
W0613:unused-argument|25
W0621:redefined-outer-name|6
W0622:redefined-builtin|8
W0631:undefined-loop-variable|2
W0640:cell-var-from-loop|1
W0702:bare-except|5
W0703:broad-except|1
W1201:logging-not-lazy|11
W1202:logging-format-interpolation|25
W1401:anomalous-backslash-in-string|1
W1501:bad-open-mode|1

**Comment by gsfr** [comment]
I have started a similar effort on scitran/reaper, which is a much smaller codebase. Let me work on that for a couple of days and then report back here what I have learned.

**Comment by gsfr** [comment]
pylintrc file updated

**Comment by gsfr** [comment]
I also added `pep8` to scitran/reaper. https://github.com/scitran/reaper/commit/7e2415c571d60a0b3c289dc62f2434dab20c1110

**Comment by ryansanford** [comment]
I pushed changes to address most of the `error` types.

@rentzso -- Can you review my changes, and comment on the correct resolution for the remaining 7? You can check results locally via `pip install -r requirements_dev.txt ; pylint --disable=W,C,R api` from the root of this repo either on your host, or within the running docker container.


```
api/files.py:125: [E0203(access-member-before-definition), getHashingFieldStorage.HashingFieldStorage._FieldStorage__write] Access to member '_FieldStorage__file' before its definition line 131
api/files.py:130: [E0203(access-member-before-definition), getHashingFieldStorage.HashingFieldStorage._FieldStorage__write] Access to member '_FieldStorage__file' before its definition line 131

api/handlers/searchhandler.py:96: [E1123(unexpected-keyword-arg), SearchHandler.get] Unexpected keyword argument '_source' in method call
api/handlers/searchhandler.py:96: [E1123(unexpected-keyword-arg), SearchHandler.get] Unexpected keyword argument 'size' in method call
api/handlers/searchhandler.py:109: [E1123(unexpected-keyword-arg), SearchHandler.get_datatree] Unexpected keyword argument 'size' in method call

api/handlers/containerhandler.py:257: [E0602(undefined-variable), ContainerHandler.get_all] Undefined variable '_id'
api/handlers/containerhandler.py:367: [E0602(undefined-variable), ContainerHandler.post] Undefined variable '_id'

**Comment by gsfr** [comment]
@ryansanford Please rebase on master and then assign to @nagem for review.

**Comment by ryansanford** [comment]
@gsfr sounds good

**Comment by ryansanford** [comment]
@nagem 

Can you review my changes, and the remaining 7 errors mentioned above?

**Comment by nagem** [comment]
Changes LGTM, fixed the few remaining errors. Looks like merging the `advanced-search` branch fixed a few as well. 

Thanks, @ryansanford!

**Comment by nagem** [comment]
Rebased, found one new error:

`api/search/queryprocessor.py:162: [E0602(undefined-variable), PreparedSearch._get_targets] Undefined variable 'min_length'`

File: api/search/queryprocessor.py [hunk 1]
```diff
@@ -159,7 +159,7 @@ def _get_targets(self, path):
                     c: [copy.deepcopy(target)] for c in self.containers
                 }
             else:
-                return {path_parts[-min_length-1]: [target]}
+                return {path_parts[-2]: [target]}
```
Comment (nagem) [inline api/search/queryprocessor.py]: @rentzso when you are available can you let me know what was intended here?

Comment (rentzso) [inline api/search/queryprocessor.py]: The implementation for search that I ported on master from st_prototype is slightly more complex (search on analyses was supported). `min_length` is used there, but not here. I just forgot to remove it...

Comment (nagem) [inline api/search/queryprocessor.py]: Ah no worries I just wasn't sure if what I filled in would have the intended functionality. Is `-2` (grabbing the container in the path above `file` or `notes`) a good enough replacement for now?

Comment (rentzso) [inline api/search/queryprocessor.py]: yes that's perfect! Thanks.

**Comment by nagem** [comment]
Remaining errors when using `elasticsearch-py`:
```
api/handlers/searchhandler.py:105: [E1123(unexpected-keyword-arg), SearchHandler.get_datatree] Unexpected keyword argument 'size' in method call
api/search/queryprocessor.py:54: [E1123(unexpected-keyword-arg), SearchContainer._exec_query] Unexpected keyword argument 'size' in method call
api/search/queryprocessor.py:128: [E1123(unexpected-keyword-arg), TargetProperty._exec_query] Unexpected keyword argument 'size' in method call
```

Looks like pylint doesn't have a good way to handle args that are injected via a decorator: https://github.com/elastic/elasticsearch-py/blob/master/elasticsearch/client/__init__.py#L456

They have a ticket open that has some recent activity:
https://github.com/PyCQA/pylint/issues/258

Until then we'll use single-line disables. We can reevaluate this plan if we add too many uses of the `elasticsearch` library before a fix is available.

**Comment by nagem** [comment]
Rebased to check `analyses` branch that was recently merged into master - this LGTM for merge as soon as CI is green!

**Comment by ryansanford** [comment]
Thanks for wrapping this up @nagem . Merging now...

**Pull request merged**